File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/headlines/NewsViewModel.java
package com.andruid.magic.newsdaily.headlines;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import androidx.lifecycle.ViewModel;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

import com.andruid.magic.newsloader.model.News;

import static com.andruid.magic.newsloader.data.Constants.PAGE_SIZE;

public class NewsViewModel extends ViewModel {
    private MutableLiveData<String> countryLiveData, categoryLiveData;
    private LiveData<PagedList<News>> pagedListLiveData;

    NewsViewModel(String country) {
        countryLiveData = new MutableLiveData<>(country);
        categoryLiveData = new MutableLiveData<>();
        PagedList.Config pagedListConfig = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(PAGE_SIZE)
                .build();
        pagedListLiveData = Transformations.switchMap(categoryLiveData, category ->
            new LivePagedListBuilder<>(new NewsDataSourceFactory(countryLiveData.getValue(),
                    categoryLiveData.getValue()), pagedListConfig).build());
    }

    public void setCountry(String country){
        countryLiveData.postValue(country);
    }

    public LiveData<PagedList<News>> getNewsForCategory(String category) {
        categoryLiveData.postValue(category);
        return pagedListLiveData;
    }
}

File: NewsDaily/app/src/main/java/com/andruid/magic/newsdaily/ui/viewmodel/NewsViewModel.kt
package com.andruid.magic.newsdaily.ui.viewmodel

import androidx.lifecycle.*
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.cachedIn
import com.andruid.magic.newsdaily.data.PAGE_SIZE
import com.andruid.magic.newsdaily.database.repository.DbRepository
import com.andruid.magic.newsloader.data.model.Result
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.flow

class NewsViewModel(selectedCountry: String, private val category: String) : ViewModel() {
    private val countryLiveData = MutableLiveData(selectedCountry)
    val newsLiveData = countryLiveData.switchMap { country ->
        flow {
            emit(Result.Loading)

            val config = PagingConfig(PAGE_SIZE)
            val pager = Pager(config) {
                DbRepository.getNews(country, category)
            }

            pager.flow
                .cachedIn(viewModelScope)
                .collect { news ->
                    emit(Result.Success(news))
                }
        }.asLiveData()
    }

    fun updateCountry(country: String) {
        countryLiveData.postValue(country)
    }
}